### ISSUE TITLE
How to train with only one class?

### ISSUE THREAD

**Issue by tiberium24** [open-mmlab/mmsegmentation#1385]
Hello,

I need to train a model which should be able to recognize an object from it's background. How do we do that?

**Comment by MengzhangLI** [comment]
DRIVE dataset has only one class, `num_classes=2`, other things you can check out from here:

https://github.com/open-mmlab/mmsegmentation/blob/master/configs/unet/fcn_unet_s5-d16_64x64_40k_drive.py

**Comment by tiberium24** [comment]
Thank you @MengzhangLI. That's exactly what I tried. For unknown reason the dice coefficient for the class I want to detect ("cell") is very low and flunctuates a lot also. The dataset is simple and easy but the dice for cell class doesn't go above 50% whereas for background it is mostly above 90%. Could something be wrong with the input? Reading png and tiff files.

**Comment by MengzhangLI** [comment]
> Thank you @MengzhangLI. That's exactly what I tried. For unknown reason the dice coefficient for the class I want to detect ("cell") is very low and flunctuates a lot also. The dataset is simple and easy but the dice for cell class doesn't go above 50% whereas for background it is mostly above 90%. Also the loss looks like:
> 
> ![image](https://user-images.githubusercontent.com/31497974/158907693-b933a1cf-9dfd-4112-806a-0c05ad448036.png)

What is ratio of cell and background? Does background take up most space of image?

**Comment by tiberium24** [comment]
Please let me know how to improve. I have been using mmdetection for years and was very satisfied with its performance so I am not sure what I am doing wrong here.

I am training with 7500 images and also testing on that same train dataset. I am getting background 97% and cells 80% dice coefficient. My training loss looks like this: 

![image](https://user-images.githubusercontent.com/31497974/159070802-0bc892c6-4010-4374-8eeb-16d254ff9711.png)

Clearly there should be a way to improve the performance. Since this is the train set that I am testing, I shouldn't be getting only 80% dice right?

Also, my labels are 0-1 for the two classes in terms of pixel values.
These are the settings I use:

- configs/unet/fcn_unet_s5-d16_64x64_40k_drive.py:

> _base_ = [
    '../_base_/models/fcn_unet_s5-d16.py', '../_base_/datasets/custom.py',
    '../_base_/default_runtime.py', '../_base_/schedules/schedule_160k.py'
]
model = dict(test_cfg=dict(crop_size=(128, 128), stride=(42, 42)))
evaluation = dict(metric='mDice')
load_from = 'checkpoints/fcn_unet_s5-d16_64x64_40k_drive_20201223_191051-5daf6d3b.pth'

- configs/_base_/models/fcn_unet_s5-d16.py:

> norm_cfg = dict(type='SyncBN', requires_grad=True)
model = dict(
    type='EncoderDecoder',
    backbone=dict(
        type='UNet',
        in_channels=3,
        base_channels=64,
        num_stages=5,
        strides=(1, 1, 1, 1, 1),
        enc_num_convs=(2, 2, 2, 2, 2),
        dec_num_convs=(2, 2, 2, 2),
        downsamples=(True, True, True, True),
        enc_dilations=(1, 1, 1, 1, 1),
        dec_dilations=(1, 1, 1, 1),
        with_cp=False,
        conv_cfg=None,
        norm_cfg=norm_cfg,
        act_cfg=dict(type='ReLU'),
        upsample_cfg=dict(type='InterpConv'),
        norm_eval=False),
    decode_head=dict(
        type='FCNHead',
        in_channels=64,
        in_index=4,
        channels=64,
        num_convs=1,
        concat_input=False,
        dropout_ratio=0.1,
        num_classes=2,
        norm_cfg=norm_cfg,
        align_corners=False,
        loss_decode=dict(
            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0, class_weight=[0.5,0.5])),
    auxiliary_head=dict(
        type='FCNHead',
        in_channels=128,
        in_index=3,
        channels=64,
        num_convs=1,
        concat_input=False,
        dropout_ratio=0.1,
        num_classes=2,
        norm_cfg=norm_cfg,
        align_corners=False,
        loss_decode=dict(
            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=0.4, class_weight=[0.5,0.5])),
    train_cfg=dict(),
    test_cfg=dict(mode='slide', crop_size=256, stride=170))

- configs/_base_/datasets/custom.py

> dataset_type = 'CustomDataset'
data_root = 'private'
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=False)
img_scale = (224, 224)
crop_size = (64, 64)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations'),
    dict(type='Resize', img_scale=img_scale, ratio_range=(0.5, 2.0)),
    dict(type='RandomCrop', crop_size=crop_size, cat_max_ratio=0.75),
    dict(type='RandomFlip', prob=0.5),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='Pad', size=crop_size, pad_val=0, seg_pad_val=255),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_semantic_seg'])
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=img_scale,
        # img_ratios=[0.5, 0.75, 1.0, 1.25, 1.5, 1.75, 2.0],
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(type='Normalize', **img_norm_cfg),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img'])
        ])
]
data = dict(
    samples_per_gpu=12,
    workers_per_gpu=0,
    train=dict(
        type=dataset_type,
        data_root=data_root,
        img_dir='train_dataset/images_png',
        ann_dir='train_dataset/masks',
        pipeline=train_pipeline),
    val=dict(
        type=dataset_type,
        data_root=data_root,
        img_dir='train_dataset/images_png',
        ann_dir='train_dataset/masks',
        pipeline=test_pipeline),
    test=dict(
        type=dataset_type,
        data_root=data_root,
        img_dir='train_dataset/images_png',
        ann_dir='train_dataset/masks',
        pipeline=test_pipeline))

- schedule_160k.py

>optimizer = dict(type='SGD', lr=0.01, momentum=0.9, weight_decay=0.0005)
optimizer_config = dict()
lr_config = dict(policy='poly', power=0.9, min_lr=1e-4, by_epoch=True)
runner = dict(type='EpochBasedRunner', max_epochs=10)
checkpoint_config = dict(by_epoch=True, interval=10)
evaluation = dict(interval=10, metric='mDice', pre_eval=True)

**Comment by MengzhangLI** [comment]
I am not sure whether it is caused by MMSegmentation itself because we never use MMSegmentation UNet to achieve competitive results in medical image segmentation tasks (2D). We would consider add medical image segmentation in the next month.

My suggestion right now: (1) If number of images samples is large, use transformer models like Swin. (2) Try to use multiple loss like Cross Entropy + Dice.

Best,

**Comment by tiberium24** [comment]
If someone finds this helpful, I figured out what was going wrong. In my input I probably did not have values of only 0 and 1 for ground truth. I made it sure it is like that by using in loading.py: 

gt_semantic_seg[gt_semantic_seg != 0] = 1

**Comment by thexray** [comment]
> I am not sure whether it is caused by MMSegmentation itself because we never use MMSegmentation UNet to achieve competitive results in medical image segmentation tasks (2D). We would consider add medical image segmentation in the next month.

@MengzhangLI Why do you never use it? What do you use instead? Is there still a plan to add medical image segmentation to MMSegmentation?
